[Kerby Soh] iP

kerbysoh opened the pull request:
# DukePro

> “First, solve the problem. Then, write the code.” – John Johnson

## Features

- Add tasks
- Mark tasks as done
- Delete tasks
- Mark tasks as undone
- Save tasks to local file
- Find tasks by name

## All you need to do is,

1. download it from [here](https://github.com/kerbysoh/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :grinning:

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

kerbysoh committed:
Level 1

kerbysoh committed:
Implement Duke Level 2

- Store text entered by the user into an array
- When requested, display all tasks in the list

kerbysoh committed:
Implement Duke Level 3

- Ability to mark classes as "done" and unmark classes as "undone"

kerbysoh committed:
Implement Duke Level 4:

- Add ToDo, Deadline and Event classes

- Add new commands to handle ToDo, Deadline and Event tasks

kerbysoh committed:
Implement use of Task Manager

- Altered code to make use of TaskManager class to handle operations such as adding tasks, marking tasks and unmarking tasks

kerbysoh committed:
Updated use of TaskManager to handle inputs

kerbysoh committed:
Automated Text UI Testing

- Usage of Input.txt and EXPECTED.TXT files to redirect input and output to semi-automate the testing of Duke functions.

kerbysoh committed:
Implement Level 5: Handle Errors

- Added DukeException and other exceptions inheriting from DukeException to handle various invalid inputs

kerbysoh committed:
Implemented Duke Level 6: Delete

- Created new function to handle deleting of tasks

kerbysoh committed:
Implement Duke Level 7

- Introduce new classes to handle parsing of commands and printing of responses

- New storage class to save user tasks locally

kerbysoh committed:
Revert "Implement Duke Level 7"

This reverts commit c2329fd607bcd578466bc6715ba7855adf1c1da3.

kerbysoh committed:
Implement Duke Level 7

- Introduce new classes to handle parsing of commands and printing of responses

- New storage class to save user tasks locally

kerbysoh committed:
Merge branch 'branch-level-7'

kerbysoh committed:
Implement Duke Level 8:

- Created a new DateParser class to handle parsing of dates, from date to string, and from string to date formats.

- Store Deadline and Event dates as java LocalDate objects, and print them in a different format.

kerbysoh committed:
Merge branch 'branch-level-8'

kerbysoh committed:
Implement packaging

- Package all files under "duke" package

kerbysoh committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

kerbysoh committed:
Implement JUnit testing using Gradle

- Setup JUnit tests using Gradle

- Created ParserFileTest to test the parsing of local storage file used to save tasks

- Created DateParserTest to test the parsing of dates used in the Deadline and Event tasks

kerbysoh committed:
Configure Gradle files to create JAR file

- Edit build.gradle and gradle-wrapper.properties to create JAR file

kerbysoh committed:
Implement additional JUnit testing

- Implement testing for parsing of tasks into strings to be written to storage

- Fix indent on switch statements

kerbysoh committed:
Implement JavaDocs

- Add JavaDocs documentation to the project

kerbysoh committed:
Implement Coding Standard and Checkstyle

- Change file to adhere to the module coding standard

- Fix Checkstyle violations

kerbysoh committed:
Implement Duke Level 9 - Find

- Search command to allow users to search for tasks matching a description

kerbysoh committed:
Merge branch 'branch-A-CodingStandard'

- Merge branch to include Coding Standard and Checkstyle

- Resolve merge conflicts

kerbysoh committed:
Merge branch 'branch-A-Level-9'

- Resolve merge conflicts

kerbysoh committed:
Update JavaDocs for Test classes

kerbysoh committed:
Merge branch 'branch-A-JavaDoc'

kerbysoh committed:
Implement search by date

- Find tasks matching a given date by using the "date" command followed by a date string

kerbysoh committed:
Implement GUI

- Added a graphical user interface to the Duke application

kerbysoh committed:
Merge branch 'GUI'

kerbysoh committed:
Assert that LocalDate object parsed into a String is not null

- DateParser::dateToString assumes that the LocalDate object is a valid LocalDate and attempts to parse it into a String

- However, passing in a null object as an argument for the function is possible and may lead to errors during the invoking of this function

- Adding an assertion to the start of this function ensures that the LocalDate object cannot be null

kerbysoh committed:
Creating a new function to format a list of tasks as a String

- Previously, the methods TaskManager::list, TaskManager::findByDate and TaskManager::find individually handles the parsing the list of tasks as a string

- This leads to code duplication as all three functions handle the same behaviour individually

- By creating a new method TaskManager:listTasks, this helps to reduce code duplication as all three functions can now call TaskManager::listTasks to concatenate the tasks as a string.

kerbysoh committed:
Use of streams to find tasks matching a description

- Previously the TaskManager:find method utilises a for loop to iterate through elements of the taskList to find those that matches the description given by the user input

- This is not as efficient as using a stream to filter the tasks

- Hence, the taskList is first converted into a stream

- Using the Stream:filter method, we can then sieve out tasks that match the given description

kerbysoh committed:
Merge pull request #1 from kerbysoh/branch-A-Assertions

Assert that LocalDate object parsed into a String is not null

kerbysoh committed:
Merge pull request #2 from kerbysoh/branch-A-CodeQuality

Creating a new function to format a list of tasks as a String

kerbysoh committed:
Merge pull request #3 from kerbysoh/branch-A-Streams

Use of streams to find tasks matching a description

kerbysoh committed:
Fix coding standard violations and improve code quality

kerbysoh committed:
Implement feature to update tasks given the index number of the task

- Remove 'final' keyword on the Task description

- Allowing the updating of task description

cwq2326 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    Deadline(String description, String deadlineDate, Boolean completed) {
```
Perhaps can rename completed to isCompleted

cwq2326 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,67 @@
+     * Obtains the type of <code>Task</code>
+     * @return String corresponding to the type of <code>Task</code>
+     */
+    @Override
+    public String getType() {
+        return "Deadline";
+    }
```
Maybe can put task type as a class attribute
```suggestion
    // Class attribute
    static final private TYPE = "Deadline";
    
    @Override
    public String getType() {
        return TYPE;
    }
```

cwq2326 commented on src/main/java/duke/TaskManager.java:
```diff
@@ -0,0 +1,184 @@
+     * @param date Date supplied by user input
+     * @return String representing tasks with given date
+     */
+    public String findByDate(String date) {
```
l like how you give the user the option to find a task by its date 👍

cwq2326 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,126 @@
+        userInput = new TextField();
+        sendButton = new Button("Send");
+
+        AnchorPane mainLayout = new AnchorPane();
+        mainLayout.getChildren().addAll(scrollPane, userInput, sendButton);
+
+        scene = new Scene(mainLayout);
+
+        stage.setScene(scene);
+        stage.show();
+
+        stage.setTitle("Duke");
+        stage.setResizable(false);
+        stage.setMinHeight(600.0);
+        stage.setMinWidth(400.0);
+
+        mainLayout.setPrefSize(400.0, 600.0);
+
+        scrollPane.setPrefSize(385, 535);
+        scrollPane.setHbarPolicy(ScrollPane.ScrollBarPolicy.NEVER);
+        scrollPane.setVbarPolicy(ScrollPane.ScrollBarPolicy.ALWAYS);
+
+        scrollPane.setVvalue(1.0);
+        scrollPane.setFitToWidth(true);
+
+        dialogContainer.setPrefHeight(Region.USE_COMPUTED_SIZE);
+
+        userInput.setPrefWidth(325.0);
+
+        sendButton.setPrefWidth(55.0);
+
+        AnchorPane.setTopAnchor(scrollPane, 1.0);
+
+        AnchorPane.setBottomAnchor(sendButton, 1.0);
+        AnchorPane.setRightAnchor(sendButton, 1.0);
+
+        AnchorPane.setLeftAnchor(userInput , 1.0);
+        AnchorPane.setBottomAnchor(userInput, 1.0);
+
+    }
```
Maybe can remove all these and let MainWindow.java handle the GUI functionalities

cwq2326 reviewed (commented):
Overall code is clean and meets coding standard. 👍 
Just some minor changes as I have suggested.

kerbysoh committed:
Implement a better GUI

- Update GUI with a prison theme

Benjykoh commented:
Overall looks good! Clean, and well written code that is easy to follow.

kerbysoh committed:
Minor changes to UI and fix coding standard violations

- Added Stage title name

- Renamed some uses of completed to isCompleted

- Removed some dead unused code

kerbysoh committed:
Set theme jekyll-theme-dinky

kerbysoh committed:
Update README.md

- Updated User Guide

kerbysoh committed:
Minor changes to User Guide

- Fix typos and formatting issues
